### PR TITLE
extract_all_metapaths: get all metapaths for any source/target metanodes

### PR DESCRIPTION
**Pull request by dhimmel**
Refs https://github.com/greenelab/hetmech/issues/76#issuecomment-378698413

Possibly relevant: code in [1-prepare.ipynb](https://github.com/dhimmel/integrate/blob/93feba1765fbcd76fd79e22f25121f5399629148/viz/auto/1-prepare.ipynb) from `dhimmel/integrate`.

### CODE REVIEW

**Comment by dhimmel** [comment]
This PR does two things:

1. Changes the existing `metagraph.extract_metapaths` function to use `source` and `target` as argument names rather than `source_kind` and `target_kind`. This function can now take metanodes or metanode_kinds for these args. Unfortunately, this could break existing code which calls these parameters by name. I don't see any code [via a GitHub search](https://github.com/search?q=extract_metapaths&type=Code) that specified these arguments by name. Now target can be set to None to get all metapaths starting on source.

2. Add `metagraph.extract_all_metapaths` which gets all metapaths below a certain length. This function has a `exclude_inverts` toggle that could help us only calculate values (like path counts) for one orientation of metapaths.

**Comment by dhimmel** [comment]
@zietzm care to review?

File: hetio/hetnet.py [hunk 1]
```diff
@@ -274,9 +274,17 @@ def add_edge(self, edge_id):
             inverse.inverted = True
             self.n_inverts += 1
 
-    def extract_metapaths(self, source_kind, target_kind, max_length):
-        source = self.node_dict[source_kind]
-        target = self.node_dict[target_kind]
+    def extract_metapaths(self, source, target=None, max_length=4):
```
Comment (zietzm) [inline hetio/hetnet.py]: Any reason we pick default `max_length=4` here but no default in `extract_all_metapaths`?

File: test/extract_metapaths_test.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    # Test max_length=0
+    metapaths = metagraph.extract_metapaths('Compound', 'Disease', max_length=0)
+    assert metapaths == []
+
+    # Test source-target-length combination where no metapaths exist
+    metapaths = metagraph.extract_metapaths('Symptom', 'Pathway', max_length=2)
+    assert metapaths == []
+
+    # Test max_length=4: the metapaths used in Project Rephetio (note that
+    # Project Rephetio excluded length 1 metapaths) resulting in 1206 metapaths.
```
Comment (zietzm) [inline test/extract_metapaths_test.py]: Could we maybe say for clarity, `1206 + 2` metapaths? I understand that Rephetio had all but length 1, so it was 1208 - 2 = 1206 but it seems like a typo on first glance.

File: test/extract_metapaths_test.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+])
+def test_extract_all_metapaths(max_length, exclude_inverts, n_metapaths):
+    """
+    Test metapath extraction on the Hetionet v1.0 metagraph. Note the expected
+    values were computed using the extract_all_metapaths function and have not
+    been independently verified.
+    """
+    metagraph = get_hetionet_metagraph()
+    metapaths = metagraph.extract_all_metapaths(max_length, exclude_inverts)
+    assert len(metapaths) == n_metapaths
```
Comment (zietzm) [inline test/extract_metapaths_test.py]: Do we need a newline at the end of the file?

File: hetio/hetnet.py [hunk 4]
```diff
@@ -293,8 +301,36 @@ def extract_metapaths(self, source_kind, target_kind, max_length):
...
+        metapaths = list()
+        metapaths_with_inverses = set()
+        metanodes = self.get_nodes()
+        for source in metanodes:
+            from_source = self.extract_metapaths(source, max_length=max_length)
+            for metapath in from_source:
+                if exclude_inverts and metapath in metapaths_with_inverses:
+                    continue
+                metapaths.append(metapath)
+                for covered in metapath, metapath.inverse:
```
Comment (zietzm) [inline hetio/hetnet.py]: Minor comment, but is there a different word than `covered` we could use here? Maybe `covered_metapath` or some synonym of covered? No big issue to just keep it though, either way.

File: hetio/hetnet.py [hunk 5]
```diff
@@ -293,8 +301,36 @@ def extract_metapaths(self, source_kind, target_kind, max_length):
...
-                     if metapath.target() == target]
+
+        if target:
+            metapaths = [metapath for metapath in metapaths
+                         if metapath.target() == target]
+        return metapaths
+
+    def extract_all_metapaths(self, max_length, exclude_inverts=False):
+        """
+        Extract all metapaths up to max_length possible by walking the
```
Comment (zietzm) [inline hetio/hetnet.py]: `Extract all possible metapaths up to max_length` might be a bit more clear.

File: test/extract_metapaths_test.py (hunk 3 above)
Comment (dhimmel) [reply test/extract_metapaths_test.py]: It should have a newline. If it didn't GitHub would display an ugly red thing.

File: test/extract_metapaths_test.py (hunk 2 above)
Comment (dhimmel) [reply test/extract_metapaths_test.py]: Ah there's a typo in that the close parenthesis should be at the end. Would that fix your confusion? Are you saying line 32 should do `1206 + 2`?

File: hetio/hetnet.py (hunk 4 above)
Comment (dhimmel) [reply hetio/hetnet.py]: Will think of something.

File: hetio/hetnet.py (hunk 1 above)
Comment (dhimmel) [reply hetio/hetnet.py]: you cannot have args without defaults come after args with defaults. I didn't want to set max_length, but had to if I wanted to set a default for target. I'm open to setting in both places if you think that's best or choosing a different default like 0 to make sure people always set it.

File: test/extract_metapaths_test.py (hunk 2 above)
Comment (zietzm) [reply test/extract_metapaths_test.py]: Moving the closing parenthesis would clear things up

File: hetio/hetnet.py (hunk 1 above)
Comment (zietzm) [reply hetio/hetnet.py]: That makes sense. Of the numbers to pick, I think 4 is  a good default.

**Pull request merged**